spring-pulsar/src/main/java/org/springframework/pulsar/config/GenericPulsarListenerEndpoint.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.pulsar.config;
+
+import org.springframework.lang.Nullable;
+
+/**
+ * Model for a Pulsar listener endpoint. Can be used against a
+ * {@link org.springframework.pulsar.annotation.PulsarListenerConfigurer} to register
+ * endpoints programmatically.
+ *
+ * @author Christophe Bornet
+ */
+public interface GenericPulsarListenerEndpoint {
+
+	@Nullable
+	String getId();
+
+	@Nullable
+	String getSubscriptionName();
```

Comment:
Also in the `ConcurrentPulsarListenerContainerFactory`.

Reply:
I mean it could be put in `PulsarListenerEndpoint` and removed from `GenericPulsarListenerEndpoint`.
Currently we need it in `GenericPulsarListenerEndpoint` because we do:
```java
                String subscriptionName = endpoint.getSubscriptionName();
		String id = endpoint.getId();

		Assert.hasText(subscriptionName, "Endpoint id must not be empty");

		synchronized (this.listenerContainers) {
			Assert.state(!this.listenerContainers.containsKey(id),
					"Another endpoint is already registered with id '" + subscriptionName + "'");
			C container = createListenerContainer(endpoint, factory);
			this.listenerContainers.put(id, container);
		}
```
in `GenericPulsarListenerEndpointRegistry`.
But I don't understand why we check `subscriptionName` and then use `id` as container key. Shouldn't we only check `id` ? I looked at spring-kafka and they only check `id`.

Reply:
We need it [here](https://github.com/spring-projects-experimental/spring-pulsar/blob/main/spring-pulsar/src/main/java/org/springframework/pulsar/config/ConcurrentPulsarListenerContainerFactory.java#L70) for the concurrent factory to extract the subscription name from the `PulsarListener` annotation provided as listener properties.

Reply:
Both `id` and `subsctiptionName` are not mandatory from a `PulsarListener` standpoint. If they are not present, the framework auto-generates them. Granted, the auto-generation of the subscription name only makes sense for an exclusive consumer; for anything else, the use might want to provide the sub name.

Reply:
With that said, we could refactor this area if there is room for further improvement.

Reply:
I am all for refactoring - can we can do it in another PR though?

Reply:
Sure